iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m

```diff
@@ -93,6 +93,92 @@ -(void)tearDown {
     [OneSignal pauseInAppMessages:false];
 }
 
+/**
+ Make sure on_session IAMs are pulled for the specific app_id
+ For this test we have mocked a single IAM in the on_session request
+ After first on_session IAMs are setup to be used by controller
+*/
+- (void)testIAMsAvailable_afterOnSession {
+    // 1. Make sure 0 IAMs are persisted
+    NSArray *cachedMessages = [OneSignalUserDefaults.initStandard getSavedCodeableDataForKey:OS_IAM_MESSAGES_ARRAY defaultValue:nil];
+    XCTAssertNil(cachedMessages);
+    
+    // 2. Open app
+    [UnitTestCommonMethods initOneSignalAndThreadWait];
+    [UnitTestCommonMethods runBackgroundThreads];
+
+    // 3. Kill the app and wait 31 seconds
+    [UnitTestCommonMethods backgroundApp];
+    [UnitTestCommonMethods runBackgroundThreads];
+    [UnitTestCommonMethods clearStateForAppRestart:self];
+    [NSDateOverrider advanceSystemTimeBy:31];
+    
+    // 4. Open app
+    [UnitTestCommonMethods initOneSignalAndThreadWait];
+    [UnitTestCommonMethods runBackgroundThreads];
+    
+    // 5. Ensure the last network call is an on_session
+    // Total calls - 2 ios params + player create + on_session = 4 requests
+    XCTAssertEqualObjects(OneSignalClientOverrider.lastUrl, serverUrlWithPath(@"players/1234/on_session"));
+    XCTAssertEqual(OneSignalClientOverrider.networkRequestCount, 4);
+    
+    // 6. Make sure IAMs are available,but not in queue
+    XCTAssertTrue([OSMessagingController.sharedInstance getInAppMessages].count > 0);
+    
+    // 7. Make sure 1 IAM is persisted
+    cachedMessages = [OneSignalUserDefaults.initStandard getSavedCodeableDataForKey:OS_IAM_MESSAGES_ARRAY defaultValue:nil];
+    XCTAssertEqual(1, cachedMessages.count);
+}
+
+/**
+ Make sure on_session IAMs are pulled for the specific app_id
+ For this test we have mocked a single IAM in the on_session request response
+ After first on_session IAMs will be cached, now force quit app and return in less than 30 seconds to make sure cached IAMs are used instead
+*/
+- (void)testIAMsCacheAvailable_afterOnSession_andAppRestart {
```

Comment:
maybe test that if the session didn't happen then IAMs shouldn't be persisted?

Reply:
My second test tests that technically

Reply:
on_session doesnt happen and the same IAMs exist in cache form before

Reply:
oh my bad I didn't look it properly